pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ServerStarterIntegrationTest.java

```diff
@@ -0,0 +1,179 @@
+import org.apache.helix.model.InstanceConfig;
+import org.apache.pinot.common.utils.CommonConstants;
+import org.apache.pinot.common.utils.ZkStarter;
+import org.apache.pinot.controller.helix.ControllerTest;
+import org.apache.pinot.server.starter.helix.HelixServerStarter;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.apache.pinot.common.utils.CommonConstants.Helix.KEY_OF_SERVER_NETTY_HOST;
+import static org.apache.pinot.common.utils.CommonConstants.Helix.KEY_OF_SERVER_NETTY_PORT;
+import static org.apache.pinot.common.utils.CommonConstants.Server.CONFIG_OF_INSTANCE_ID;
+import static org.apache.pinot.common.utils.CommonConstants.Server.CONFIG_OF_USE_LOGICAL_INSTANCE_ID;
+import static org.testng.Assert.assertEquals;
+
+
+public class ServerStarterIntegrationTest extends ControllerTest {
+  public static final String DEFAULT_SERVER_ID = "Server_127.0.0.1_8098";
+  public static final String SERVER1 = "Server1";
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    startZk();
+    startController();
+  }
+
+  @AfterClass
+  public void tearDown()
+      throws Exception {
+    stopController();
+    stopZk();
+  }
+
+  @Test
+  public void testWithNoInstanceIdNoHostnamePort()
+      throws Exception {
+    // Test the behavior when no instance id nor hostname/port is specified in server conf.
+    Configuration serverConf = new PropertiesConfiguration();
+    // Start the server
+    HelixServerStarter helixServerStarter =
+        new HelixServerStarter(getHelixClusterName(), ZkStarter.DEFAULT_ZK_STR, serverConf);
+
+    // Verify the serverId, host and port are set correctly in Zk.
+    HelixManager helixManager = helixServerStarter.getHelixManager();
+    PropertyKey.Builder keyBuilder = helixManager.getHelixDataAccessor().keyBuilder();
+    InstanceConfig config =  helixManager.getHelixDataAccessor().getProperty(keyBuilder.instanceConfig(DEFAULT_SERVER_ID));
+    helixServerStarter.stop();
+
+    assertEquals(config.getInstanceName(), DEFAULT_SERVER_ID);
+    // By default (auto joined instances), server instance name is of format: {@code Server_<hostname>_<port>}, e.g.
+    // {@code Server_localhost_12345}, hostname is of format: {@code Server_<hostname>}, e.g. {@code Server_localhost}.
+    // More details refer to the class ServerInstance.
+    assertEquals(config.getHostName(), "Server_127.0.0.1");
+    assertEquals(config.getPort(), "8098");
+  }
+
+  @Test
+  public void testWithNoInstanceIdButWithHostnamePort()
+      throws Exception {
+    // Test the behavior when no instance id specified but hostname/port is specified in server conf.
+    Configuration serverConf = new PropertiesConfiguration();
+    serverConf.setProperty(KEY_OF_SERVER_NETTY_HOST, "host1");
+    serverConf.setProperty(KEY_OF_SERVER_NETTY_PORT, 10001);
+    // Start the server
+    HelixServerStarter helixServerStarter =
```

Comment:
Following code looks like its repeated in every test. Consider refactoring ?

Reply:
Code refactored to remove duplicate codes.